Capital:
- 2,000,000 shells

Goal:
- Maximize shells

Constraints:
- Maximum of 5 trades
- Start and end with shells
- Each trade involves trading all available capital (can't hold >1 currency at any given time)

In [2]:
# 0: pizza, 1: wasabi, 2: snowball, 3: shells
rates = [[1, 0.48, 1.52, 0.71],
         [2.05, 1, 3.26, 1.56],
         [0.64, 0.3, 1, 0.46],
         [1.41, 0.61, 2.08, 1]]

Brute force approach
- try all possible combinations of trades that start and end with shells
- shells -> p,w,sb,s -> p,w,sb,s -> p,w,sb,s -> p,w,sb,s -> shells
    - 4^4 = 256 possibilities
- allow 1:1 trades (as placeholders for blank trades) as optimal solution may involve <5 trades

In [3]:
def dfs(curr, amt, trades, solns):
    if len(trades)==5:
        # trade to shells and append soln
        solns.append(trades + [3] + [amt * rates[curr][3]])
    else:
        for i in range(4):
            solns = dfs(i, amt*rates[curr][i], trades+[i], solns)
    return solns

In [4]:
solns = dfs(3, 2000000, [3], [])
print(len(solns)) # confirm all solutions computed

256


In [5]:
max_sol = max(solns, key=lambda x: x[-1])
print(max_sol) # optimal soln (hopefully)

[3, 0, 1, 3, 0, 3, 2113938.7776]


In [6]:
amt = 2000000
for i in range(1,len(max_sol)-1):
    print(f"{max_sol[i-1]} -> {max_sol[i]}: {amt} -> {amt*rates[max_sol[i-1]][max_sol[i]]}")
    amt *= rates[max_sol[i-1]][max_sol[i]]

3 -> 0: 2000000 -> 2820000.0
0 -> 1: 2820000.0 -> 1353600.0
1 -> 3: 1353600.0 -> 2111616.0
3 -> 0: 2111616.0 -> 2977378.56
0 -> 3: 2977378.56 -> 2113938.7776


The below 2 trades seem to be the most profitable. A constraint of 5 trades allows us to perform each strategy once.

shells -> pizza -> shells
shells -> pizza -> wasabi -> shells

### Optimal Solns:
- shells -> pizza -> wasabi -> shells -> pizza -> shells
- shells -> pizza -> shells -> pizza -> wasabi -> shells

Shells: 2,000,000 -> 2,113,938.7776

### Submission:

shells -> pizza -> wasabi -> shells -> pizza -> shells

In [7]:
# print all possible trade sequences
for s in solns:
    print(s)

[3, 0, 0, 0, 0, 3, 2002200.0]
[3, 0, 0, 0, 1, 3, 2111616.0]
[3, 0, 0, 0, 2, 3, 1971744.0]
[3, 0, 0, 0, 3, 3, 2002200.0]
[3, 0, 0, 1, 0, 3, 1970164.7999999996]
[3, 0, 0, 1, 1, 3, 2111616.0]
[3, 0, 0, 1, 2, 3, 2029858.56]
[3, 0, 0, 1, 3, 3, 2111616.0]
[3, 0, 0, 2, 0, 3, 1947740.16]
[3, 0, 0, 2, 1, 3, 2006035.2]
[3, 0, 0, 2, 2, 3, 1971744.0]
[3, 0, 0, 2, 3, 3, 1971744.0]
[3, 0, 0, 3, 0, 3, 2004402.42]
[3, 0, 0, 3, 1, 3, 1905293.52]
[3, 0, 0, 3, 2, 3, 1915704.9600000002]
[3, 0, 0, 3, 3, 3, 2002200.0]
[3, 0, 1, 0, 0, 3, 1970164.7999999996]
[3, 0, 1, 0, 1, 3, 2077830.1439999996]
[3, 0, 1, 0, 2, 3, 1940196.096]
[3, 0, 1, 0, 3, 3, 1970164.7999999996]
[3, 0, 1, 1, 0, 3, 1970164.7999999996]
[3, 0, 1, 1, 1, 3, 2111616.0]
[3, 0, 1, 1, 2, 3, 2029858.56]
[3, 0, 1, 1, 3, 3, 2111616.0]
[3, 0, 1, 2, 0, 3, 2005147.2384]
[3, 0, 1, 2, 1, 3, 2065160.448]
[3, 0, 1, 2, 2, 3, 2029858.56]
[3, 0, 1, 2, 3, 3, 2029858.56]
[3, 0, 1, 3, 0, 3, 2113938.7776]
[3, 0, 1, 3, 1, 3, 2009413.7856]
[3, 0, 1, 3, 2, 3, 2020394